pkg/downloader/chart_downloader.go

```diff
@@ -92,25 +92,28 @@ func (c *ChartDownloader) DownloadTo(ref, version, dest string) (string, *proven
 		return "", nil, err
 	}
 
-	g, err := c.Getters.ByScheme(u.Scheme)
-	if err != nil {
-		return "", nil, err
-	}
-
-	data, err := g.Get(u.String(), c.Options...)
-	if err != nil {
-		return "", nil, err
-	}
-
 	name := filepath.Base(u.Path)
 	if u.Scheme == registry.OCIScheme {
 		idx := strings.LastIndexByte(name, ':')
 		name = fmt.Sprintf("%s-%s.tgz", name[:idx], name[idx+1:])
 	}
+	g, err := c.Getters.ByScheme(u.Scheme)
+	if err != nil {
+		return "", nil, err
+	}
 
 	destfile := filepath.Join(dest, name)
-	if err := fileutil.AtomicWriteFile(destfile, data, 0644); err != nil {
-		return destfile, nil, err
+
+	_, err = os.Stat(destfile)
+	if err != nil {
+		data, err := g.Get(u.String(), c.Options...)
```

Comment:
What if you have two repos with a chart of the same name and version but their content is different?

The existing functionality handles that by downloading the latest. The new functionality would cause an issue but not report anything.

For a cache, we might need something that's content based (e.g., hash based). Name version pairs is not a safe cache key.

Reply:
> What if you have two repos with a chart of the same name and version but their content is different?

Isn't this a design problem with Helm's cache? I'm not calculating cache locations myself, and the way they are calculated allows for this scenario, i.e. if you happened to use multiple repos with the same chart name and version, they would use the same cache location. Right now the system works correctly only because Helm doesn't use the cache at all. If it did, this scenario would fail immediately and especially provenance checks would fail badly.

> For a cache, we might need something that's content based (e.g., hash based). Name version pairs is not a safe cache key.

Certainly, but this patch (and also #12628) builds upon existing cache functionality, reworking how the cache is set up would be a much larger undertaking. If that's desired, we could go that way, which would probably among others entail using ETags and similar techniques to verify the download.

Reply:
> Isn't this a design problem with Helm's cache?

Yes. You are correct. Before the cache can be used the design problem needs to be fixed.

We are open to bigger changes to correct the design.

Reply:
my understanding is that the existing behaviour and the user expectations that are based on it are prohibitive for a cache. users would publish different variations of charts using the same version and expect them to be picked up in subsequent helm invocations. Maybe an opt-in toggle would be an option `--use-cache`? 

Here would be a broader discussion around this topic: https://github.com/helm/community/pull/185